File: business/src/test/java/com/pij/lottieshow/interactor/DirectorySourceTest.java
package com.pij.lottieshow.interactor;

import com.pij.lottieshow.model.LottieFile;

import org.junit.Rule;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;

import java.io.File;

import rx.observers.TestSubscriber;

import static java.util.Collections.emptyList;
import static java.util.Collections.singletonList;
import static org.mockito.Mockito.when;
import static rx.Observable.just;
import static rx.Observable.never;

/**
 * <p>Created on 10/04/2017.</p>
 * @author Pierrejean
 */
public class DirectorySourceTest {

    @Rule
    public MockitoRule mockito = MockitoJUnit.rule();
    @Mock
    File mockRoot;

    @Test
    public void emitsEmptyListWhenRootDoesNotEmit() {
        when(mockRoot.isDirectory()).thenReturn(false);
        DirectorySource sut = new DirectorySource(never());
        TestSubscriber<Iterable<LottieFile>> subscriber = TestSubscriber.create();

        sut.lottieFiles().subscribe(subscriber);

        subscriber.assertNoErrors();
        subscriber.assertValue(emptyList());
    }

    @Test
    public void emitsEmptyListWhenRootIsNotDirectory() {
        when(mockRoot.isDirectory()).thenReturn(false);
        DirectorySource sut = createDefaultSut(mockRoot);
        TestSubscriber<Iterable<LottieFile>> subscriber = TestSubscriber.create();

        sut.lottieFiles().subscribe(subscriber);

        subscriber.assertNoErrors();
        subscriber.assertValue(emptyList());
    }

    @Test
    public void emitsEmptyListWhenDirectoryEmpty() {
        when(mockRoot.isDirectory()).thenReturn(true);
        when(mockRoot.listFiles()).thenReturn(new File[0]);
        DirectorySource sut = createDefaultSut(mockRoot);
        TestSubscriber<Iterable<LottieFile>> subscriber = TestSubscriber.create();

        sut.lottieFiles().subscribe(subscriber);

        subscriber.assertNoErrors();
        subscriber.assertValue(emptyList());
    }

    @Test
    public void emitsSingletonListWhenDirectoryContainsOneFile() {
        when(mockRoot.isDirectory()).thenReturn(true);
        when(mockRoot.listFiles()).thenReturn(new File[]{ new File("zip", "zap") });
        DirectorySource sut = createDefaultSut(mockRoot);
        TestSubscriber<Iterable<LottieFile>> subscriber = TestSubscriber.create();

        sut.lottieFiles().subscribe(subscriber);

        subscriber.assertNoErrors();
        //noinspection unchecked
        subscriber.assertValues(emptyList(), singletonList(LottieFile.create(new File("zip", "zap"))));
    }

    private DirectorySource createDefaultSut(File root) {
        return new DirectorySource(just(root));
    }
}

File: business/src/test/java/com/pij/lottieshow/interactor/DirectorySourceTest.kt
package com.pij.lottieshow.interactor

import com.pij.lottieshow.model.LottieFile
import org.junit.Rule
import org.junit.Test
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.junit.MockitoJUnit
import rx.Observable.just
import rx.Observable.never
import rx.observers.TestSubscriber
import java.io.File
import java.util.Collections.emptyList

/**
 *
 * Created on 10/04/2017.
 * @author Pierrejean
 */
class DirectorySourceTest {

    @get:Rule var mockito = MockitoJUnit.rule()
    @Mock internal lateinit var mockRoot: File

    @Test fun `Emits empty list when root does not emit`() {
        val sut = DirectorySource(never<File>())
        val subscriber = TestSubscriber.create<Iterable<LottieFile>>()

        sut.lottieFiles().subscribe(subscriber)

        subscriber.assertNoErrors()
        subscriber.assertValue(listOf())
    }

    @Test fun `Emits empty list when root is not directory`() {
        `when`(mockRoot.isDirectory).thenReturn(false)
        val sut = createDefaultSut(mockRoot)
        val subscriber = TestSubscriber.create<Iterable<LottieFile>>()

        sut.lottieFiles().subscribe(subscriber)

        subscriber.assertNoErrors()
        subscriber.assertValue(listOf())
    }

    @Test fun `Emits empty list when root is an empty directory`() {
        `when`(mockRoot.isDirectory).thenReturn(true)
        `when`(mockRoot.listFiles()).thenReturn(arrayOfNulls<File>(0))
        val sut = createDefaultSut(mockRoot)
        val subscriber = TestSubscriber.create<Iterable<LottieFile>>()

        sut.lottieFiles().subscribe(subscriber)

        subscriber.assertNoErrors()
        subscriber.assertValue(listOf())
    }

    @Test fun `Emits singleton list when root contains a single file`() {
        `when`(mockRoot.isDirectory).thenReturn(true)
        `when`(mockRoot.listFiles()).thenReturn(arrayOf(File("zip", "zap")))
        val sut = createDefaultSut(mockRoot)
        val subscriber = TestSubscriber.create<Iterable<LottieFile>>()

        sut.lottieFiles().subscribe(subscriber)

        subscriber.assertNoErrors()

        subscriber.assertValues(emptyList<LottieFile>(), listOf(LottieFile.create(File("zip", "zap"))))
    }

    private fun createDefaultSut(root: File): DirectorySource {
        return DirectorySource(just(root))
    }
}